src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+
+	// max possible length 32: "Confirm x of x"
+	// where x is 2^32-1
+	char text[33];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d of %d", ctx->nativeScript.nOfK.n, ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - N of K",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	VALIDATE(expectsRequest(), ERR_INVALID_STATE);
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
+	ctx->level++;
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native complex script type = %u", nativeScriptType);
+
+	switch(nativeScriptType) {
+#	define  CASE(TYPE, HANDLER) case TYPE: HANDLER(view); break;
+		CASE(NATIVE_SCRIPT_ALL, deriveNativeScriptHash_handleAll);
+		CASE(NATIVE_SCRIPT_ANY, deriveNativeScriptHash_handleAny);
+		CASE(NATIVE_SCRIPT_N_OF_K, deriveNativeScriptHash_handleNofK);
+		default: THROW(ERR_INVALID_DATA);
+#	undef   CASE
+	}
+}
+
+// Simple native scripts
+
+void deriveNativeScriptHash_handleDeviceOwnedPubkey(read_view_t* view)
+{
+	bip44_path_t *path = &ctx->nativeScript.pubkeyDeviceOwned.path;
+	view_skipBytes(view, bip44_parseFromWire(path, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view)));
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("Pubkey given by path:");
+	BIP44_PRINTF(path);
+	PRINTF("\n");
+
+	ui_displayPathScreen(
+		"Script - key path",
+		&ctx->nativeScript.pubkeyDeviceOwned.path,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleThirdPartyPubkey(read_view_t* view)
+{
+	uint8_t *hash = ctx->nativeScript.pubkeyThirdParty.hash;
+	VALIDATE(view_remainingSize(view) == ADDRESS_KEY_HASH_LENGTH, ERR_INVALID_DATA);
+	view_memmove(hash, view, ADDRESS_KEY_HASH_LENGTH);
+
+#	ifdef DEVEL
+	TRACE("Pubkey given by hash:");
+	for (int i = 0; i < ADDRESS_KEY_HASH_LENGTH; i++) {
+		PRINTF("%x", hash[i]);
+	}
+	PRINTF("\n");
+#	endif
+
+	ui_displayHexBufferScreen(
+		"Script - key",
+		ctx->nativeScript.pubkeyThirdParty.hash,
+		ADDRESS_KEY_HASH_LENGTH,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handlePubkey(read_view_t* view)
+{
+	uint8_t pubkeyType = parse_u1be(view);
+	TRACE("pubkey type = %u", pubkeyType);
+
+	switch(pubkeyType) {
+#	define  CASE(TYPE, HANDLER) case TYPE: HANDLER(view); break;
+		CASE(PUBKEY_DEVICE_OWNED, deriveNativeScriptHash_handleDeviceOwnedPubkey);
+		CASE(PUBKEY_THIRD_PARTY, deriveNativeScriptHash_handleThirdPartyPubkey);
+		// any other value for the pubkey type is invalid
+		default: THROW(ERR_INVALID_DATA);
+#	undef   CASE
+	}
+}
+
+#define SPLIT_UINT64(n) (uint32_t) (n >> 32), (uint32_t) n
```

Comment:
Variables should be always parenthesized in macro definitions.

But I'm not sure if we need this splitting at all. I guess I would prefer `TRACE_UINT64` here.